lib/ex_aws/s3.ex

```diff
@@ -517,19 +517,21 @@ defmodule ExAws.S3 do
   |> ExAws.request! #=> :done
   ```
   """
-  def upload(source, bucket, path, opts \\ []) do
-    do_upload(source, bucket, path, opts)
-  end
-
-  defp do_upload(%Flow{} = flow, bucket, path, opts) do
+  @spec upload(
+    source :: Enumerable.t | Flow.t,
+    bucket :: String.t,
+    path :: String.t,
+    opts :: Keyword.t) :: __MODULE__.Upload.t
```

Comment:
Here's what I ended up with for a typespec. I saw that ExDoc only displays one typespec, and checked in Ecto to see how they handle an optional keyword list: https://github.com/elixir-ecto/ecto/blob/v2.1.0-rc.0/lib/ecto/changeset.ex#L1065. I also went with `String.t` for the bucket and path instead of binary because I think we want a standard UTF-8 encoded string not a possibly non-printable binary.

Let me know if you prefer the typespec how you previously had it.